Support for TSQL UNPIVOT Operator

manisha-deshpande opened the pull request:
### Description

The UNPIVOT operator in T-SQL is a powerful relation operator for transforming data from a wide format (columns) into a long format (rows). It rotates columns of a table-valued expression into column values, converting multiple columns into row values, under two new columns: 
* A dimension column containing the original column names
* A measure column containing the values from those columns

```
TABLE:
customer_turnover( customer_id INT, q1 INT, q2 INT, q3 INT, q4 INT)

QUERY:
SELECT customer_id, turnover, quarter
FROM customer_turnover
UNPIVOT (
    turnover    -- measure column (will contain values)
    FOR quarter -- dimension column (will contain column names)
    IN (        -- list of source columns to unpivot
        q1, q2, q3, q4
    )
) AS unpvt; 
```

This PR implements TSQL UNPIVOT operator support in Babelfish by transforming it into equivalent PostgreSQL CROSS JOIN LATERAL operations. Currently, Babelfish does not support UNPIVOT operations. With this change, users can now use TSQL UNPIVOT syntax to transform wide-format data into long format.

UNPIVOT is a crucial TSQL feature used extensively in data transformation and analytics. Supporting this feature improves compatibility for migrating SQL Server applications to Babelfish.

The implementation adds TSQL UNPIVOT support through a multi-stage transformation approach. 
- At the parser stage, new grammar rules recognize TSQL UNPIVOT syntax and transform it into PostgreSQL's CROSS JOIN LATERAL structure by creating a JoinExpr node. This transformation also handles source table aliasing (if alias not present), and reassign the unpivot alias as the complete Join alias, returning the transformed node in a list with other metadata (unpivot identifier, unpivot alias, newly produced measure column etc). 
- At the analyzer stage, the unpivot hook function recursively processes the SelectStmt's fromClause tree to find and unwrap unpivot "List" nodes to make the query tree valid, whilst processing the other meta data from the node to create WHERE IS NOT NULL clause for the dimension column(s) (NULL exclusion (matching SQL Server's behavior)), and to handle SELECT * expansion by filtering out unpivot source columns.

### Issues Resolved

BABEL-4307: Support for TSQL UNPIVOT operators
Support for CTE, VIEW, JOIN

### Test Scenarios Covered ###
* **Use case based -**
* Basic unpivot syntax with target list variations
```
SELECT customer_id, turnover, quarter 
FROM customer_turnover 
UNPIVOT (turnover FOR quarter IN (q1, q2, q3, q4)) AS unpvt;
```
* JOINS
```
SELECT c.customer_name, u.turnover, u.quarter 
FROM customer_info c
JOIN customer_turnover t
UNPIVOT (turnover FOR quarter IN (q1, q2, q3, q4)) AS u
ON c.customer_id = t.customer_id;
```
* VIEWS
* CTE
* DML Operations (INSERT, UPDATE)
* Temporary objects (Variables, Temp Tables, Derived tables)
* Support numeric, string, date datatypes

* **Boundary conditions -**
* Maximum length of table, unpivot column names
* text with 4000 characters

* **Arbitrary inputs -**
* Special characters
* NULL values

* **Performance tests -**
* Large dataset: 1K rows, 100K rows

* **Other tests -**
TBD



### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
Basic UNPIVOT parse rules and transformation func

Initial commit that accepts basic unpivot statement and extracts
required information in the transformation function in the pre analyzer

Task: BABEL-4307
Signed-off-by: “manisha-deshpande” <“dmanisha.work@gmail.com”>

Commit:
Readjusted unpivot grammar rules, transformation

Moved unpivot rule from simple_select clause to table_ref clause to accept consecutive unpivots. Added transformation logic to transform unpivot stmt to cross join lateral stmt, except alias generation functionality and updated target list

Task: BABEL-4307
Signed-off-by: “manisha-deshpande” <“dmanisha.work@gmail.com”>

Commit:
Added target aliasing and where condition logic

Passed unpivot context such as "UNPIVOT" marker, unpivot table alias and
its two column names, source table/query alias that is newly generated
to simple_select to perform target alias update and adding NOT_NULL where clause
for dimension column

Task: BABEL-4307
Signed-off-by: “manisha-deshpande” <“dmanisha.work@gmail.com”>

Commit:
Added unpivot hook for alias and NULL handling

Moved target alias updation and WHERE clause NULL handling logic from parser
to analyzer stage by introducing unpivot hook function

Task: BABEL-4307
Signed-off-by: “manisha-deshpande” <“dmanisha.work@gmail.com”>

Commit:
Added recursive dfs logic to update unpivot node

Signed-off-by: “manisha-deshpande” <“dmanisha.work@gmail.com”>

Commit:
Consecutive unpivots and source alias generator

Added support/fix for consecutive unpivots by considering unpivot list
nodes as source. Added a new function that generates a new alias for
source table/query of unpivot, if not present.

Task: BABEL-4307
Signed-off-by: “manisha-deshpande” <“dmanisha.work@gmail.com”>

Commit:
Added alias handling

Introduced outer alias for JoinExpr inorder to bring the source table
and cross join subquery under one alias. This helps tackle the TSQL
unpivot expected vs Postgres alias mismatch while referencing columns in
target list/ other parts of the query.

Task: BABEL-4307
Signed-off-by: “manisha-deshpande” <“dmanisha.work@gmail.com”>

Commit:
Added fix for SELECT * issue

Added function filter_star_targetlist_for_unpivot that executes
parse_sub_analyze on the statement if * is present in targetlist.
This function iterates the analyzed targetlist and creates a new
targetlist for the current stmt with the unpivot source columns skipped.

Task: BABEL-4307
Signed-off-by: “manisha-deshpande” <“dmanisha.work@gmail.com”>

Commit:
Minor changes and additional unpivot test cases

Task: BABEL-4307
Signed-off-by: “manisha-deshpande” <“dmanisha.work@gmail.com”>

Commit:
fixed spacing issues

manisha-deshpande commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2016,6 +2025,238 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+		JoinExpr *join = (JoinExpr *)table_ref;
+		Node *last_table = join->rarg;
+		
+		if (IsA(last_table, RangeVar))
```
repeated logic

Commit:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into jira-babel-4307
Merging upstream into current branch

manisha-deshpande committed:
Additional unpivot tests

Added tests for SET operations, DML operations (INSERT/UPDATE/DELETE),
subqueries, table valued function  with unpivot

Signed-off-by: “manisha-deshpande” <dmanisha.work@gmail.com>

manisha-deshpande committed:
Added expected output files for tests

Signed-off-by: “manisha-deshpande” <dmanisha.work@gmail.com>

manisha-deshpande commented on test/JDBC/expected/unpivot-vu-cleanup.out:
might be incomplete. needs revision.

manisha-deshpande committed:
Added support for functin, json sources

Moved the logic related to aliasing based on the source type to a new
function get_unpivot_source_alias(Node *table_ref) with if else cases.
Introduced two more types: RangeFunction, RangeTableFunc. Update the
test files with the examples of said types(Function, json).

Signed-off-by: “manisha-deshpande” <dmanisha.work@gmail.com>

manisha-deshpande committed:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into jira-babel-4307

manisha-deshpande committed:
Added boundary condition tests and fixed output files

Signed-off-by: “manisha-deshpande” <dmanisha.work@gmail.com>

forestkeeper commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -1820,7 +1820,11 @@ TsqlForXMLMakeFuncCall(TSQL_ForClause *forclause)
 	fc = makeFuncCall(func_name, func_args, COERCE_EXPLICIT_CALL, -1);
 
 	rt->name = palloc0(4);
+	//CWE-676,120 - Insecure Buffer Access -HIGH VULN
```
Can you explain a little for this comment in here ?

manisha-deshpande replied:
Tried a fix for a vulnerability issue unrelated to unpivot. Removed.

forestkeeper commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -1903,7 +1907,12 @@ TsqlForJSONMakeFuncCall(TSQL_ForClause *forclause)
 	rt->name = palloc0(5);
+	// CWE-676,120 - Insecure Buffer Access -HIGH VULN
 	strncpy(rt->name, "json", 4);
+	//Instead, use
```
Why we add this comment ?

forestkeeper commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2077,253 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+	else if (IsA(table_ref, JoinExpr))
+	{
+		/* For JOIN, we need to ensure the last joined table has an alias */
+		/* TODO: First extract Join Alias if present else rarg */
```
normally we don't put TODO in the comment, if we do, pls create another jira to track in here.

manisha-deshpande replied:
agreed. Followed up on TODO comment and found an issue with "grouped" joins as unpivot source. Documented in known issues.

Edit: Resolved issue by removing the previous code which picked the join expression's right argument for unpivot alias.
Fix: Pick the grouped's join's alias if present, else created. Tests worked.

forestkeeper commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2077,253 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+        
+        ereport(ERROR,
+                (errcode(ERRCODE_SYNTAX_ERROR),
+                 errmsg("Invalid JOIN expression in UNPIVOT source")));
```
Where is the test case for this ?

forestkeeper commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2077,253 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+	n->isNatural = false;
+	//source_alias = "invalid_default_alias"; /* Will be overwritten or error thrown */ 
+	/* Set up left side with alias if not present */
+	source_alias = get_unpivot_source_alias(table_ref);
```
Will table_ref == NULL ?

manisha-deshpande replied:
The parse rule for table_ref ensures it is not NULL (`gram.y`). SQL with empty table_ref will be stopped by a syntax error at the parser stage itself.

manisha-deshpande replied:
Added an assert in the beginning of `get_unpivot_source_alias` function itself.

forestkeeper commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2077,253 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+    	RangeTableFunc *rtf = (RangeTableFunc *)table_ref;
+		if (rtf->alias == NULL)
+		{
+			rtf->alias = makeAlias(generate_unpivot_source_table_alias("json"), NIL);
```
What's the purpose for making alias ?

manisha-deshpande replied:
For TSQL unpivot, the source table does not require an alias.
However after a transformation to cross join lateral, the values list requires an alias to reference from (since lateral allows direct referencing from the left source table), i.e  
`table1 UNPIVOT (a for b in (q1, q2)) as u` essentially transforms to
`table1 T CROSS JOIN LATERAL (VALUES (T.q1, 'q1'), (T.q2, 'q2')) u(a, b)`

So a custom logic to add an alias if missing, is required.

forestkeeper commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2077,253 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+	if(source_alias == NULL){
+		ereport(ERROR,
+			(errcode(ERRCODE_SYNTAX_ERROR),
+			errmsg("Invalid source structure for UNPIVOT operation")));
```
Where is the test case for this error msg

manisha-deshpande replied:
Looked and tested various unpivot SQLs to reach this error, couldn't find any. 

Background: 
"UNPIVOT relational operator can be used to change a TABLE-VALUED expression into another TABLE" is the generic definition for unpivot from sql server documentation.
Referencing the potential from-clauses in `postgres_modified_for_bbf/backend/parser/transformFromClause -> transformFromClauseItem`, the code covers all cases except RangeTableSample. Therefore, unpivot would fail here if the source is not one of:
- `RangeVar` (regular tables)
- `RangeSubselect` (proper subqueries within FROM)
- `JoinExpr` (proper JOIN operations)
- `RangeFunction` (table-valued functions)
- `RangeTableFunc` (JSON/XML functions)
- `List` (previous UNPIVOT operations)

The error message is a precautionary handler incase an incorrect structure passes through table_ref in the unpivot_clause parse rule.
Please let me know if it should still be removed.

manisha-deshpande replied:
Removed. Moved error report to end of `get_unpivot_source_alias` function, as part of other [pr comment](https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/3431#discussion_r1994037997)

forestkeeper commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2077,253 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+	/* Build VALUES list from unpivot source columns */
+	foreach(lc, source_cols)
+	{
+		// TODO: needs to be llast if aliases are used in values list, ex c.q1
```
Why there's a TODO in here ?

manisha-deshpande replied:
Issue cannot be fixed here in code, more relevant to parser
Add error test case

forestkeeper commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2077,253 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+		col_ref = makeNode(ColumnRef);
+		col_ref->fields = list_make2(
+			makeString(source_alias), /* source table alias */
+			makeString(strVal(col_name)) /* column name */
```
col_name is already a String, why have to do this ? Is it for copying a string ?

forestkeeper commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2077,253 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+	{
+		ereport(ERROR,
+					(errcode(ERRCODE_SYNTAX_ERROR),
+					 errmsg("Alias for UNPIVOT operation is required.")));
```
Where is the test case for this error msg

manisha-deshpande replied:
Removed error statement and preceding null check. Unpivot Alias does not require a null check as the grammar rule "alias_clause" will ensure it is not NULL, otherwise error is thrown.
Added related test case.

manisha-deshpande replied:
Not null check enforced in parse rule
Add an assert instead here with explanation.

forestkeeper commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-prologue.y.h:
```diff
@@ -18,6 +18,11 @@
+// #include "nodes/nodes.h"
+// #include "nodes/parsenodes.h"
+// #include "nodes/value.h"
+// #include "parser/parse_node.h"
```
Delete those lines

forestkeeper commented on contrib/babelfishpg_tsql/src/tsqlUnsupportedFeatureHandler.cpp:
```diff
@@ -1404,8 +1404,8 @@ antlrcpp::Any TsqlUnsupportedFeatureHandlerImpl::visitCheckpoint_statement(TSqlP
-	if (ctx->UNPIVOT())
-		handle(INSTR_UNSUPPORTED_TSQL_UNPIVOT, ctx->UNPIVOT());
+	// if (ctx->UNPIVOT())
+	// 	handle(INSTR_UNSUPPORTED_TSQL_UNPIVOT, ctx->UNPIVOT());
```
Should delete the whole function, not just the function body

forestkeeper commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5346,6 +5354,194 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+
+	/* Free allocated memory */
+	list_free_deep(measure_cols);
+	//list_free_deep(src_cols);
```
delete those lines

forestkeeper commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5346,6 +5354,194 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+			found_unpivot = true;
+			found_unpivot |= transform_unpivot_clause_recursive(node_ptr, where_clause, measure_cols, unpivot_src_cols);
+		}
+	}
```
Is there any other possibilities ?

forestkeeper replied:
If there's no other possible for this if clause, pls add an Assert here.

manisha-deshpande replied:
This dfs function simply looks for "UNPIVOT" list nodes and if found, executes the unwrapping of the list to appropriate node structure and recurses on it to find deeper unpivot list nodes. 
The `if` case checks for a list starting with the string "UNPIVOT".
The `else` case would be to do nothing and return found_unpivot=false.

manisha-deshpande replied:
Add comment explaining else case: ignore other cases except join and List
Returns boolean if unpivot node is found.

forestkeeper commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5346,6 +5354,194 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+	ListCell *lc;
+	
+	/* Only process if target list contains * */
+	// TODO: Validate agains all types of target list (including functions)
```
left //TODO

manisha-deshpande replied:
Is a future enhancement, updated comment. Already tracked in known list of bugs.
Issue: Can handle `Select * ...`, Cannot handle `Select unpivot_alias.* ...`
Current behavior: gives results but with extra columns (i.e, filter_star_targetlist_for_unpivot is not executed)

Edit tracked in: BABEL-5677

manisha-deshpande replied:
Suggestion:
Test and add cases for more variations in filter_star_targetlist_for_unpivot
COUNT(\*), SUM(\*)
table1.\*, table2.\*
'\*'
"\*"
table1.'\*'
table1.'\*'
'table1.\*'

forestkeeper commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5346,6 +5354,194 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+		unpivot_info = (List *)*node_ptr;
+		if (list_length(unpivot_info) == 7 &&
+			IsA(linitial(unpivot_info), String) &&
+			strcmp(strVal(linitial(unpivot_info)), "UNPIVOT") == 0)
```
What about lower case/upper case ?

manisha-deshpande replied:
This check does not check the unpivot key word. The string is set by us in the unpivot_transformation function during parsing to highlight that this is an unpivot node and checked later to unwrap the list into proper structure.

manisha-deshpande committed:
Resolve pr comments

Removed unrelated vuln fix comment. Removed unused include library
statements. added some error condition test cases. Removed null check
for unpivot alias as grammar rule ensures it is not null.

Signed-off-by: “manisha-deshpande” <dmanisha.work@gmail.com>

manisha-deshpande committed:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into jira-babel-4307

manisha-deshpande committed:
Addresed Pr comments, remove unintended changes

Signed-off-by: “manisha-deshpande” <dmanisha.work@gmail.com>

manisha-deshpande committed:
Addressed pr comment

Removed unecessary makeString(strVal()) call on col_name as it is
already a (String *) type. However it is essential while using
makeStringConst, throws error otherwise.

Signed-off-by: “manisha-deshpande” <dmanisha.work@gmail.com>

RIC06X commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-prologue.y.h:
```diff
@@ -56,6 +56,9 @@ static Node * buildTsqlMultiLineTvfNode(int create_loc, bool replace, List *func
 										char *tokens_remaining, int tokens_loc, bool alter, core_yyscan_t yyscanner);
 static Node *tsql_pivot_select_transformation(List *target_list, List *from_clause, List *pivot_clause, Alias *alias_clause, SelectStmt *pivot_sl);
+static Node *tsql_unpivot_transformation(List *components);
+char *generate_unpivot_source_table_alias(const char *base);
```
we don't we use static here?

manisha-deshpande committed:
Fixed test, added procedure and dynamic unpivotsql

Signed-off-by: “manisha-deshpande” <dmanisha.work@gmail.com>

RIC06X commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2068,249 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+			rtf->alias = makeAlias(generate_unpivot_source_table_alias("json"), NIL);
+		}
+		return rtf->alias->aliasname;
+	}
```
Check if we need to handle `RangeTableSample`, I saw this one is mentioned in t`ransformFromClauseItem`

manisha-deshpande replied:
RangeTableSample not currently supported by BBF.
Add assert

RIC06X commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2068,249 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+	{
+		/* Handle case where table_ref is an UNPIVOT node */
+		List *prev_unpivot = (List *)table_ref;
+		if (list_length(prev_unpivot) == 7 &&
```
Add NULL check for `prev_unpivot` List

manisha-deshpande committed:
Fixed bug for joined table as unpivot source

Updated join logic in generate_unpivot_source_table_alias to use its own
alias if exists, or to create one. Previous logic was incorrect and
assumed the alias of the right argument of the join. Also updated test
case output, currently failing for lowered case result. Also updated the
logic of filter_star_targetlist_for_unpivot to not filter out column
twice (duplicate).

Task: BABEL-4307
Signed-off-by: “manisha-deshpande” <dmanisha.work@gmail.com>

manisha-deshpande commented on test/JDBC/expected/unpivot-vu-verify.out:
```diff
@@ -0,0 +1,1381 @@
+GO
+~~START~~
+int#!#numeric#!#varchar
+1#!#2.00#!#Q1_売上
```
NOTE:
Failing test cases: unpivot on `[Sales$Data@2024]` and `[Global_データ_Sales]`. 
```
Expected: 1#!#2.00#!#Q1_売上
Output:   1#!#2.00#!#q1_??
```
Investigating.. Would appreciate any insights.

manisha-deshpande replied:
Critical Issue: If the columns inside the unpivot list (source columns: q1, q2..) start with upper case,
unpivot is not maintaining their upper case property when using them in the result set values under the new dimension column. 
```
Ex:
Input: table has capitalised column name Q1

table unpivot( .. for Quarter in (Q1) )

Output: 
Quarter
_______
q1
```

Edit: The lower capitalization is supposedly a known issue, since postgres treats column as case insensitive

manisha-deshpande replied:
Tracked in BABEL-5676. Added relevant test case as well.

manisha-deshpande committed:
Added null check, pr comment

Signed-off-by: “manisha-deshpande” <dmanisha.work@gmail.com>

manisha-deshpande committed:
Updated failing test case result

Signed-off-by: “manisha-deshpande” <dmanisha.work@gmail.com>

manisha-deshpande committed:
edited schedule file

Signed-off-by: “manisha-deshpande” <dmanisha.work@gmail.com>

manisha-deshpande committed:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into jira-babel-4307

manisha-deshpande commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2068,244 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+		}
+
+	}
+	return NULL;
```
Add assert, shouldn't reach here.

manisha-deshpande replied:
Added ereport instead of assert. @forestkeeper 
Reason: An assert leads to `TCP Provider: Error code 0x2746 Communication link failure` whereas an error report would provide more information on the invalid UNPIVOT statement.

manisha-deshpande commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5324,6 +5332,200 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+	List *result_targetlist = NIL;
+	ListCell *lc;
+	
+	/* Only process if target list contains * 
```
Test and add cases for more variations in filter_star_targetlist_for_unpivot
COUNT(*), SUM(*)
table1.*, table2.*
'*'
"*"
table1.'*'
table1.'*'
'table1.*'

manisha-deshpande committed:
Addresed PR comments

In get_unpivot_source_alias parser function, added assert to ensure
table_ref is NON-NULL and error repoert for unlisted unpivot source
objects. Made parser functions static. Added test case for unpivot on
grouped join. Added error test cases for tracked known issues
BABEL-5676, BABEL-5677

Task: BABEL-4307
Signed-off-by: “manisha-deshpande” <dmanisha.work@gmail.com>

manisha-deshpande commented on test/JDBC/input/unpivot-vu-verify.sql:
```diff
@@ -0,0 +1,836 @@
+GO
+
+        -- 2.2 Views - Create view with UNPIVOT
+CREATE VIEW sales.sales_analysis_view AS
```
Create view test already covered. @forestkeeper

manisha-deshpande committed:
commented collation test

Non-english character test with unpivot has conflicting results in
collation and non-collation environments in automated pr tests.
Commenting test ase and errorneous result is tracked in BABEL-5676

Task: BABEL-4307
Signed-off-by: “manisha-deshpande” <dmanisha.work@gmail.com>

manisha-deshpande committed:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into jira-babel-4307

Signed-off-by: “manisha-deshpande” <dmanisha.work@gmail.com>

manisha-deshpande committed:
Commented unused SQL

Task: BABEL-4307
Signed-off-by: “manisha-deshpande” <dmanisha.work@gmail.com>

manisha-deshpande committed:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into jira-babel-4307

forestkeeper commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5324,6 +5332,200 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+static void 
+transform_unpivot_clause(ParseState *pstate, SelectStmt *stmt)
+{
+	Node *where_clause = stmt->whereClause;
```
Have to add a sql_dialect check in here, because in this pr : https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/534/files 
didn't check sql_dialect.

manisha-deshpande replied:
Added check

manisha-deshpande committed:
Added SQL dialect check in unpivot hook

Task: BABEL-4307
Signed-off-by: “manisha-deshpande” <dmanisha.work@gmail.com>

forestkeeper committed:
Merge branch 'BABEL_5_X_DEV' into jira-babel-4307

RIC06X commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2068,249 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+ * Ensures:
+ * - Unique alias per query
+ * - Length within NAMEDATALEN (base + 7 chars)
+ */
```
Please follow the function header pattern:
* Starts with a concise summary of what this function does.
* Follows by parameter list comments
* Follows by a Note section (optional) on more detailed context, and nuances.
* Each section is separated by a blank line.

Example:
```
/*
 * Get a list of file names matching the input pattern
 *
 * to_matches: a string for pattern to match
 * matchFiles: output of a list of matched file names
 * match_count: output of matched file counts
 * 
 * return: 1 if operation succeeded, 0 otherwise
 *
 * Note: this function incurs IO and may cause performance issues if repeated called
 */
static int 
get_file_matches(char* to_matches, char** matchedFiles, int* match_count) 
{
}
```

RIC06X commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2068,249 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+	
+	/* Use "src" as base if none provided */
+	if (!base)
+	{	base = "src"; }
```
If we only have 1 line in for the if stmt, then we will not use `{}`.

RIC06X commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2068,249 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+	/* Ensure base name length fits within NAMEDATALEN with suffix */
+	base_len = strlen(base);
+	if (base_len + 7 > NAMEDATALEN) /* 7 = underscore + 6 digits */
+	{	base_len = NAMEDATALEN - 7; }
```
If we only have 1 line in for the if stmt, then we will not use {}.

RIC06X replied:
Please use a constant for 7 within this function. Also move comment after the constant definition.

RIC06X commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2068,249 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+	return result;
+}
+
+/* Helper function to get the alias of the unpivot source
```
Please follow the pg style comment, first line should be empty
```
/*
 * comment text begins here
 * and continues here
 */
```

RIC06X commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2068,249 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+
+	Assert(table_ref != NULL);
+
+    if (IsA(table_ref, RangeVar))
```
Please replace all tab with 4 spaces. The if stmt here is using 4 spaces, but for the Assert above, it is using tab

RIC06X commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2068,249 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+	List *result_info;
+	char *measure_colname;
+	char *dim_colname;
+	char *inner_alias_name;
```
Please reorder this variable declaration list to group variables with same type together.

RIC06X commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2068,249 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+	char *inner_alias_name;
+	
+	/* Extract components */
+	table_ref = (Node *)linitial(components);
```
Add an assertion for `components` to make sure the input parameter is not NULL.

RIC06X commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2068,249 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+	foreach(lc, source_cols)
+	{
+		String *col_name = (String *)lfirst(lc);
+		/* Future enhancement: Handle aliased unpivot source columns syntax
```
Please follow the pg style comment, first line should be empty
```
/*
 * comment text begins here
 * and continues here
 */
```
And also ADD a TODO in the comments: `/* TODO [JIRA]: JIRA description */`

RIC06X commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2068,249 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+
+	n->larg = table_ref;
+	
+	/* Build VALUES list from unpivot source columns */
```
I recommend add the following to your PR description:
```
SELECT customer_id, turnover, quarter
FROM customer_turnover
UNPIVOT (
    turnover    -- measure column (will contain values)
    FOR quarter -- dimension column (will contain column names)
    IN (        -- list of source columns to unpivot
        q1, q2, q3, q4
    )
) AS unpvt;      
```
This help explain what is source column, measure column and dimention column

RIC06X commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2068,249 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+		col_ref->fields = list_make2(
+			makeString(source_alias), /* source table alias */
+			col_name /* column name */
+		);
```
Please update the format here:
```
/* Create pair (source table alias, column name) */
col_ref->fields = list_make2(makeString(source_alias), 
                             col_name);
```
or
```
col_ref->fields = list_make2(makeString(source_alias), col_name);
```
If the line length does not exceed 80 chars, then we will use 1 line instead.
https://www.postgresql.org/docs/current/source-format.html
"Limit line lengths so that the code is readable in an 80-column window. (This doesn't mean that you must never go past 80 columns. For instance, breaking a long error message string in arbitrary places just to keep the code within 80 columns is probably not a net gain in readability.)"

RIC06X commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2068,249 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+	values_subquery->valuesLists = values_list;
+	rarg->subquery = (Node *)values_subquery;
+	
+	/* Handle alias for Join-Values (RangeSubSelect) clause 
```
comment style here

RIC06X commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2068,249 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+		makeString(measure_colname),
+		makeString(dim_colname)
+		)
+	);
```
format:
```
	rarg->alias = makeAlias(inner_alias_name, 
							list_make2(makeString(measure_colname),
									   makeString(dim_colname))
						    );
```

RIC06X commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2068,249 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+	}
+	/* Future addition: RangeTableSample, when it is supported. */
+	ereport(ERROR,
+		(errcode(ERRCODE_SYNTAX_ERROR),
```
please fix the indentation here

RIC06X commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2068,249 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+		makeString(dim_colname),			/* dimension column */
+		makeString(measure_colname),		/* measure column */
+		makeString(source_alias)			/* source table/query alias */
+	);
```
need to revise the format here

RIC06X commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -2068,6 +2068,249 @@ tsql_pivot_select_transformation(List *target_list, List *from_clause, List *piv
+		 * Tracked in BABEL-5677
+		 */
+		List *value_pair;
+		ColumnRef *col_ref;
```
please move all the variable declaration to the start of this code block

RIC06X commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5324,6 +5332,202 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+static void 
+transform_unpivot_clause(ParseState *pstate, SelectStmt *stmt)
+{
+	if (sql_dialect == SQL_DIALECT_TSQL) {
```
Please have the following instead.
```
if (sql_dialect != SQL_DIALECT_TSQL)
    return;
```

RIC06X commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5324,6 +5332,202 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+		ListCell *lc;
+		List *measure_cols = NIL;
+		List *src_cols = NIL;
+		bool has_unpivot = false;
```
move the variable declarations on the start of the code block

RIC06X commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5324,6 +5332,202 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+ *
+ * Returns: true if UNPIVOT found and processed
+ */
+static bool transform_unpivot_clause_recursive(Node **node_ptr, Node **where_clause, List **measure_cols, List **unpivot_src_cols)
```
revise function declaration format to 
```
static bool
transform_unpivot_clause_recursive(Node **node_ptr, Node **where_clause, List **measure_cols, List **unpivot_src_cols)
```

RIC06X replied:
Also can we simply use `Node *node_ptr`?

RIC06X replied:
can you update all the other parameters? I don't think we need a double pointer here

manisha-deshpande replied:
https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/3431#discussion_r2004572848

Similarly we need to pass the reference of a list to function if the function is modifying the list. Replacing the double pointer with single `List *` is leading to failing test cases.

RIC06X commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5324,6 +5332,202 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+ * Returns: Filtered target list excluding unpivot source columns
+ * Note: Only processes if target list contains * 
+ */
+static List * filter_star_targetlist_for_unpivot(ParseState *pstate, SelectStmt *stmt, List *source_cols)
```
```
static List *
filter_star_targetlist_for_unpivot(ParseState *pstate, SelectStmt *stmt, List *source_cols)
```

RIC06X commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5324,6 +5332,202 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
 }
 
+static void 
+transform_unpivot_clause(ParseState *pstate, SelectStmt *stmt)
```
Since all stmt with from clause will run through this function. It is better to add a description on why we need to do like this.

RIC06X commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5324,6 +5332,202 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+
+		foreach(lc, stmt->fromClause)
+		{
+			has_unpivot |= transform_unpivot_clause_recursive((Node**)&(lc->ptr_value), &where_clause, &measure_cols, &src_cols);
```
can we use lfirst here:
```
has_unpivot |= transform_unpivot_clause_recursive((Node**)&lfirst(lc), &where_clause, &measure_cols, &src_cols);
```
Also should we use `Node **` here? Can we use `Node *`?

manisha-deshpande replied:
Good catch. 
However, we need `Node**` for node_ptr because we modify the original tree structure by replacing UNPIVOT List nodes with their transformed `JoinExpr` nodes.
This requires changing the pointer value itself (`*node_ptr = transformed_node`), not just the node contents.

RIC06X commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5324,6 +5332,202 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+	if (IsA(*node_ptr, JoinExpr))
+	{
+		join = (JoinExpr *)*node_ptr;
+		found_unpivot |= transform_unpivot_clause_recursive(&join->larg, where_clause, measure_cols, unpivot_src_cols);
```
if we use `Node *node_ptr`, then we can simply this line as:
```
found_unpivot |= transform_unpivot_clause_recursive(join->larg, where_clause, measure_cols, unpivot_src_cols);
```

manisha-deshpande replied:
We need `Node**` for node_ptr because we modify the original tree structure by replacing UNPIVOT List nodes with their transformed `JoinExpr` nodes.
This requires changing the pointer value itself (`*node_ptr = transformed_node`), not just the node contents.

manisha-deshpande replied:
Same as https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/3431#discussion_r2004572848

RIC06X commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5324,6 +5332,202 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+static List * filter_star_targetlist_for_unpivot(ParseState *pstate, SelectStmt *stmt, List *source_cols)
+{
+	Query *temp_query;
+	List *result_targetlist = NIL;
```
Please separate the variable assignment with variable declaration. variable declaration first

RIC06X commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5324,6 +5332,202 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+			has_unpivot |= transform_unpivot_clause_recursive((Node**)&(lc->ptr_value), &where_clause, &measure_cols, &src_cols);
+		}
+
+		if (has_unpivot)
```
Please refrain from using nested if-else.
we can do something like the following to have a early return if clean up is needed:
```
int func(char* input)
{
    char* buffer = NULL;

    buffer = malloc(100);
    if (input == NULL) 
    {
        /* error, go to clean*/
        goto CLEANUP;
    }

    /* processing data*/
    while(true)
    {
        if (done)
            goto CLEANUP;

        if (error)
            goto CLEANUP;
    }

CLEANUP:
    if (buffer != NULL)
        free(buffer);
    
```

RIC06X commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5324,6 +5332,202 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+			/* Replace UNPIVOT info with transformed node and recurse on it */
+			*node_ptr = transformed_node;
+			found_unpivot = true;
+			found_unpivot |= transform_unpivot_clause_recursive(node_ptr, where_clause, measure_cols, unpivot_src_cols);
```
Is it necessary to have `|=` here, we already set `found_unpivot = true` before, so found_unpivot will always return true in this line.

manisha-deshpande replied:
The found unpivot list node has a representative `JoinExpr` (`transformed_node`) node whose left/right arguments (subtrees) may contain `unpivot` List Nodes further down. (Example situation: consecutive unpivots)

Therefore the `unpivot JoinExpr`'s `larg` and `rarg` need to be traversed and checked for unpivot nodes again.

RIC06X replied:
`found_unpivot = true | transform_unpivot_clause_recursive(node_ptr...)` will always return true, so it is not necessary to have `|=` here.

RIC06X commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5324,6 +5332,202 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+	/* Only process if target list contains * 
+	 * Does not handle: `SELECT unpivot_alias.* ...`
+	 * Future enhancement: Validate agains more targetlist variations
+	 */
```
fix comment style here, and please also create a JIRA and add a TODO here

RIC06X commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5324,6 +5332,202 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+	{
+		TargetEntry *te = (TargetEntry *)lfirst(lc);
+		bool skip_column = false;
+	    ListCell *source_lc;
```
Variable declaration should before variable value assignment

RIC06X commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5324,6 +5332,202 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+	    ListCell *source_lc;
+
+		/* Check if this column is in source_cols */
+		for (source_lc = list_head(source_cols); source_lc != NULL;)
```
can we use `foreach` here?

manisha-deshpande replied:
Since modifying the source list while iterating with `foreach` can be unsafe and lead to undefined behavior; have used traditional `for-loop` to check for and subsequently delete matched column cell from list.

manisha-deshpande replied:
resolved: https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/3431#discussion_r2006543551

manisha-deshpande committed:
Fixed formatting issues

Task: BABEL-4307
Signed-off-by: “manisha-deshpande” <dmanisha.work@gmail.com>

manisha-deshpande committed:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into jira-babel-4307

manisha-deshpande committed:
Fixed indentation tab to spaces

Signed-off-by: “manisha-deshpande” <dmanisha.work@gmail.com>

manisha-deshpande committed:
Addressed PR comment

Removed nested ifs and introduced cleanup process and modified if checks
Removed unused parameter whereclause from
transform_unpivot_clause_recursive.

Task: BABEL-4307
Signed-off-by: “manisha-deshpande” <dmanisha.work@gmail.com>

manisha-deshpande committed:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into jira-babel-4307

RIC06X commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5338,201 +5338,247 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
-		list_free_deep(measure_cols);
-	}
+    if (measure_cols == NIL)
+        goto cleanup;
```
can we simply return here? if `measure_cols` is `NIL` and then we do a `list_free_deep`, will that cause a double free issue?

RIC06X commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5330,6 +5338,248 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+            {
+                skip_column = true;
+                /* Remove the matched source column to avoid duplicate removal */
+                source_cols = list_delete_cell(source_cols, source_lc);
```
you can use `foreach` with `foreach_delete_current`, this can guarantee deletion safety
```
/*
 * foreach_delete_current -
 *	  delete the current list element from the List associated with a
 *	  surrounding foreach() loop, returning the new List pointer.
 *
 * This is equivalent to list_delete_cell(), but it also adjusts the foreach
 * loop's state so that no list elements will be missed.  Do not delete
 * elements from an active foreach loop's list in any other way!
 */
#define foreach_delete_current(lst, cell)	\
	(cell##__state.i--, \
	 (List *) (cell##__state.l = list_delete_cell(lst, cell)))
```

manisha-deshpande committed:
Fixed PR comment

Replaced traditional for with enhanced  foreach loop and made use of foreach_delete_current

Signed-off-by: “manisha-deshpande” <dmanisha.work@gmail.com>

manisha-deshpande committed:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into jira-babel-4307

manisha-deshpande committed:
Fixing static code analyzer warning

the list being updated has to be passed by reference if it's to be
updated and used outside it's for loop. Updated list argument to list
pointer arg.

Signed-off-by: “manisha-deshpande” <dmanisha.work@gmail.com>

RIC06X commented on test/JDBC/input/unpivot-vu-verify.sql:
```diff
@@ -0,0 +1,837 @@
+FROM customer_turnover 
+UNPIVOT (turnover FOR quarter IN (q1, q2, q3, q4)) AS unpvt  
+WHERE turnover > 200;
+        
```
missing `GO` here

RIC06X commented on test/JDBC/input/unpivot-vu-verify.sql:
```diff
@@ -0,0 +1,837 @@
+WHERE CAST(turnover AS DECIMAL(10,2)) > 150.00 
+    AND quarter IN ('q1', 'q2') 
+    AND unpvt.customer_id < 1000;
+
```
`GO`

RIC06X commented on test/JDBC/input/unpivot-vu-verify.sql:
```diff
@@ -0,0 +1,837 @@
+UNPIVOT (turnover FOR quarter IN (q1, q2, q3, q4)) AS unpvt 
+ORDER BY turnover DESC 
+OFFSET 5 ROWS FETCH NEXT 5 ROWS ONLY;
+
```
add `go` to the above stmt

RIC06X commented on test/JDBC/expected/unpivot-vu-verify.out:
```diff
@@ -0,0 +1,1410 @@
+FROM customer_turnover 
+UNPIVOT (turnover FOR quarter IN (q1, q2, q3, q4)) AS unpvt  
+WHERE turnover > 200;
+        
```
missing output, need to add `GO` to all stmts, check `GO` in verify.sql file

manisha-deshpande replied:
Missed GO statements for multiple WHERE, ORDER, GROUP BY statements. Added them all and verified outputs with SQL Server.

manisha-deshpande committed:
Added missing GO statements to vu-verify tests

Task: BABEL-4307
Signed-off-by: “manisha-deshpande” <dmanisha.work@gmail.com>

manisha-deshpande committed:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into jira-babel-4307

manisha-deshpande committed:
Editing comment

Signed-off-by: “manisha-deshpande” <dmanisha.work@gmail.com>
